debugging.md

```diff
@@ -0,0 +1,107 @@
+# Debugging
+
+This document aims to collect info on the current status and future opportunities for debugging Rust.
+
+## Debugging architecture overview
+
+The compiler generates debuginfo for Rust programs (see https://github.com/rust-lang/rust/tree/master/src/librustc_trans/debuginfo).
+This is an LLVM-specific thing. LLVM then processes this through optimisations
+(with varying degrees of success) and outputs debug symbols with the library.
+These are DWARF for the Linux, Mac, and Windows/GNU toolchains, and pdb (program
+database) for Windows/visual studio (pdb support is WIP in LLVM, I believe).
+
+Debuggers consume the generated symbols. Visual Studio only works on Windows.
+LLDB works everywhere but is primarily used as a backend for XCode on MacOS. GDB
+works everywhere. GDB and LLDB are command
+line tools, but can have frontends.
+
+Debugger support can come in various forms in this architecture:
+* we could invent our own debugger and debug symbols (this is infeasible due to resourcing)
+* generate better debuginfo in the compiler - most of the low hanging fruit here is done
+* extend debug symbols with Rust-specific forms
```

Comment:
The problem is that we would like to *change* the DWARF representation of some things, not just extend it in a backwards compatible way. As one example, how we represent enums at the moment is really a hack and I personally would like to get rid of it. At the same time, I would prefer if we didn't break existing debugger versions out there until there's solid support for the new format.

My current thinking is to add a compiler flag, `-Zrust-dwarf` or something, that allow to opt into the new format, and start making changes in LLVM, GDB, and LLDB as needed. The debuginfo generated here would only be expected to work with our branches of debuggers initially. At the same time the old format is still emitted by default until we are ready to flip the switch.

I'd be very interested though in hearing about alternative strategies.